### ISSUE TITLE
[keymgr] questions about LC disable

### ISSUE THREAD

**Issue by weicaiyang** [lowRISC/opentitan#4680]
@tjaychen 

I have a few questions about LC disabling keymgr
> If the key manager has not been initialized, it cannot be initialized until life cycle enables key manager.

For this statement in the spec, keymgr can't be initialized is due to cfgen=0 if keymgr_en isn't On, right?

> Once the system settles to Disabled state, the behavior is consistent with normal behavior.

This statement seems not precise. Since cfgen is 0 when keymgr_en isn't On, all operations will be ignored after the system settles to Disabled state. Is it worth adding the cfgen is set to 0 in the spec?

From below codes, if state is in Disabled and keymgr_en is set to off, we don't enter `StCtrlWipe` to wipe secret. In this case, sideload key will be valid unless SW generates another one to override it. SW_SHARE will be still available if SW doesn't read it to clear. Will it be a risk?
```
       StCtrlWipe: begin
        stage_sel_o = Disable;
        op_done_o = op_start_i & kmac_done_i;
        op_accepted = op_start_i;
        wipe_key_o = 1'b1;

        for (int i = 0; i < Shares; i++) begin
          key_state_d[i] = {EntropyRounds{entropy_i}};
        end

        if (!op_start_i) begin
          state_d = StCtrlDisabled;
        end
      end

      // Terminal state (StDisabled is included)
      // However, it will continue to kick off dummy transactions
      default: begin
        op_done_o = op_start_i & kmac_done_i;

        // accept any command, but always select fake data
        op_accepted = op_start_i;
        stage_sel_o = Disable;

        // During disabled state, continue to update state
        key_state_d = (op_done_o && advance_sel) ? kmac_data_i : key_state_q;

        invalid_op = op_start_i;
      end
```

**Comment by tjaychen** [comment]
> @tjaychen
> 
> I have a few questions about LC disabling keymgr
> 
> > If the key manager has not been initialized, it cannot be initialized until life cycle enables key manager.
> 
> For this statement in the spec, keymgr can't be initialized is due to cfgen=0 if keymgr_en isn't On, right?
> 
yes that and the `keymgr_ctrl` fsm itself. If the enable is low, it also will not allow us to transition out of reset state.
So hopefully the protection is two-fold.

> > Once the system settles to Disabled state, the behavior is consistent with normal behavior.
> 
> This statement seems not precise. Since cfgen is 0 when keymgr_en isn't On, all operations will be ignored after the system settles to Disabled state. Is it worth adding the cfgen is set to 0 in the spec?
> 

so to clarify, `Disabled` state and keymgr_en are kind of orthogonal.
Because it is possible for software to issue a transition into disable state also with keymgr_en high the entire time. 
Does that make sense? 

> From below codes, if state is in Disabled and keymgr_en is set to off, we don't enter `StCtrlWipe` to wipe secret. In this case, sideload key will be valid unless SW generates another one to override it. SW_SHARE will be still available if SW doesn't read it to clear. Will it be a risk?

if we are already in Disabled, we are okay. There are 2 ways to transition to diasabled.
1. software directly issues a command -> in this case, it is up to software whether it wants to wipe the sideload keys
2. keymgr_en went low after being high.  In this case, we go through and wipe, and everything is gone.

Does this help? Maybe I can make this clearer in the spec. 

> 
> ```
>        StCtrlWipe: begin
>         stage_sel_o = Disable;
>         op_done_o = op_start_i & kmac_done_i;
>         op_accepted = op_start_i;
>         wipe_key_o = 1'b1;
> 
>         for (int i = 0; i < Shares; i++) begin
>           key_state_d[i] = {EntropyRounds{entropy_i}};
>         end
> 
>         if (!op_start_i) begin
>           state_d = StCtrlDisabled;
>         end
>       end
> 
>       // Terminal state (StDisabled is included)
>       // However, it will continue to kick off dummy transactions
>       default: begin
>         op_done_o = op_start_i & kmac_done_i;
> 
>         // accept any command, but always select fake data
>         op_accepted = op_start_i;
>         stage_sel_o = Disable;
> 
>         // During disabled state, continue to update state
>         key_state_d = (op_done_o && advance_sel) ? kmac_data_i : key_state_q;
> 
>         invalid_op = op_start_i;
>       end
> ```

**Comment by weicaiyang** [comment]
> so to clarify, Disabled state and keymgr_en are kind of orthogonal.
Because it is possible for software to issue a transition into disable state also with keymgr_en high the entire time.
Does that make sense?

After enter disabled state, I will also test some operations.
However, if disabled state is entered by keymgr_en=Off, the behavior is not really consistent with being entered by SW. I feel the statement "Once the system settles to Disabled state, the behavior is consistent with normal behavior" may be a bit inaccurate. Perhaps we can mention keymgr_en=Off will set cfgen=0 in spec, so that we know any operation will be ignored.


> if we are already in Disabled, we are okay. There are 2 ways to transition to diasabled.
> 
> software directly issues a command -> in this case, it is up to software whether it wants to wipe the sideload keys
> keymgr_en went low after being high. In this case, we go through and wipe, and everything is gone.
> Does this help? Maybe I can make this clearer in the spec.

I tried to test all secrets (hw key, sw_share_output) will be wiped if keymgr_en turns 0. Then I found this case:
SW enters disabled without clearing sideload key and sw_share_output, then keymgr_en turns low, sideload key and sw_share_output won't be wiped. In this case, we need SW manually wipe the secrets.

I'm also thinking a case. hackers may make SW enter Disabled state to bypass HW wipe secret logic, then try to steal HW sideload key.

**Comment by tjaychen** [comment]
> > so to clarify, Disabled state and keymgr_en are kind of orthogonal.
> > Because it is possible for software to issue a transition into disable state also with keymgr_en high the entire time.
> > Does that make sense?
> 
> After enter disabled state, I will also test some operations.
> However, if disabled state is entered by keymgr_en=Off, the behavior is not really consistent with being entered by SW. I feel the statement "Once the system settles to Disabled state, the behavior is consistent with normal behavior" may be a bit inaccurate. Perhaps we can mention keymgr_en=Off will set cfgen=0 in spec, so that we know any operation will be ignored.

o i see what you're saying. Yes i can do that.  Let me explain that. 

> 
> > if we are already in Disabled, we are okay. There are 2 ways to transition to diasabled.
> > software directly issues a command -> in this case, it is up to software whether it wants to wipe the sideload keys
> > keymgr_en went low after being high. In this case, we go through and wipe, and everything is gone.
> > Does this help? Maybe I can make this clearer in the spec.
> 
> I tried to test all secrets (hw key, sw_share_output) will be wiped if keymgr_en turns 0. Then I found this case:
> SW enters disabled without clearing sideload key and sw_share_output, then keymgr_en turns low, sideload key and sw_share_output won't be wiped. In this case, we need SW manually wipe the secrets.
> 
> I'm also thinking a case. hackers may make SW enter Disabled state to bypass HW wipe secret logic, then try to steal HW sideload key.

hmmmm... then if that's the concern we could separate Disabled from an actual invalid state.

**Comment by tjaychen** [comment]
thinking about this more, even if i'm not concerned about it, making the behavior more consistent will probably be easier for DV and future use.

So let's do this.. i'll create another state called `StCtrlInvalid`.  This will be the state we go to whenever `keymgr_en` goes low, even if we were in `StCtrlDisabled`.  To get to this new state, we must always go through a control wipe, but otherwise, its behavior will be identical to disabled.  How does that sound? 

This should then make all the behavior consistent.

**Comment by weicaiyang** [comment]
> thinking about this more, even if i'm not concerned about it, making the behavior more consistent will probably be easier for DV and future use.
> 
> So let's do this.. i'll create another state called `StCtrlInvalid`. This will be the state we go to whenever `keymgr_en` goes low, even if we were in `StCtrlDisabled`. To get to this new state, we must always go through a control wipe, but otherwise, its behavior will be identical to disabled. How does that sound?
> 
> This should then make all the behavior consistent.

sounds good to me. It's certainly more consistent for DV. But do we need to change the operation `Disable` to `Invalid` too?

**Comment by tjaychen** [comment]
no, because i think it's worth having a state where we can't do anything
"new", but it preserves the last snapshot..until it's overridden.
It's not super usable IMO, but you never know.

**Comment by weicaiyang** [comment]
> no, because i think it's worth having a state where we can't do anything "new", but it preserves the last snapshot..until it's overridden. It's not super usable IMO, but you never know.

I see. Sounds good.

**Comment by weicaiyang** [comment]
addressed via #4755
